From 863ef95ecfca1615674b5469ed5e601e73f7337d Mon Sep 17 00:00:00 2001
From: Nicola Guerrera <guerrera.nicola@gmail.com>
Date: Thu, 23 Oct 2025 19:41:44 +0200
Subject: [PATCH 73/77] power: supply: add idtp9418 driver

This is the orignial driver from xiaomi bsp-open that has been severely cleaned up and ported to mainline linux. It currently implements only basic features:

- pen detection and charging
- reporting of the state of charge
- naive charge limiting

It has been only tested on xiaomi-nabu, but it should also work for other xiaomi devices that use this charger.
---
 .../bindings/power/supply/idt,p9418.yaml      |  77 ++
 .../boot/dts/qcom/sm8150-xiaomi-nabu.dts      |  78 +-
 arch/arm64/configs/sm8150.config              |   1 +
 drivers/power/supply/Kconfig                  |  10 +
 drivers/power/supply/Makefile                 |   3 +-
 drivers/power/supply/idtp9418.c               | 794 ++++++++++++++++++
 drivers/power/supply/idtp9418.h               | 208 +++++
 7 files changed, 1169 insertions(+), 2 deletions(-)
 create mode 100644 Documentation/devicetree/bindings/power/supply/idt,p9418.yaml
 create mode 100644 drivers/power/supply/idtp9418.c
 create mode 100644 drivers/power/supply/idtp9418.h

diff --git a/Documentation/devicetree/bindings/power/supply/idt,p9418.yaml b/Documentation/devicetree/bindings/power/supply/idt,p9418.yaml
new file mode 100644
index 000000000000..576345c37bdf
--- /dev/null
+++ b/Documentation/devicetree/bindings/power/supply/idt,p9418.yaml
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: GPL-2.0
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/power/supply/idt,p9418.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: IDT P9418 Wireless Power Charger
+
+maintainers:
+  - Viola Guerrera <viooo@anche.no>
+
+properties:
+  compatible:
+    const: idt,p9418
+
+  reg:
+    maxItems: 1
+    description: I2C address of the device
+
+  irq-gpios:
+    maxItems: 1
+    description: |
+      GPIO used as the interrupt line from the chip to the host.
+
+  pen-det-active-high-gpios:
+    maxItems: 1
+    description: |
+      GPIO used to detect pen presence (active high).
+
+  pen-det-active-low-gpios:
+    maxItems: 1
+    description: |
+      GPIO used to detect pen presence (active low).
+
+  enable-gpios:
+    maxItems: 1
+    description: |
+      GPIO used to enable or disable the wireless power transmitter chip.
+
+  boost-enable-gpios:
+    maxItems: 1
+    description: |
+      GPIO used to enable the boost converter for reverse charging.
+
+  '#address-cells':
+    const: 1
+
+  '#size-cells':
+    const: 0
+
+required:
+  - compatible
+  - reg
+  - enable-gpios
+  - boost-enable-gpios
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+
+    i2c17: i2c@17 {
+        status = "okay";
+
+        idtp9418: idtp9418@3b {
+            compatible = "idt,p9418";
+            reg = <0x3b>;
+            status = "okay";
+            irq-gpios = <&tlmm 95 GPIO_ACTIVE_LOW>;
+            pen-det-active-high-gpios = <&tlmm 101 GPIO_ACTIVE_HIGH>;
+            pen-det-active-low-gpios = <&tlmm 134 GPIO_ACTIVE_HIGH>;
+            enable-gpios = <&tlmm 3 GPIO_ACTIVE_HIGH>;
+            boost-enable-gpios = <&tlmm 47 GPIO_ACTIVE_HIGH>;
+        };
+    };
+...
diff --git a/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts b/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts
index 9adf4f00254b..dfebc7f3e70c 100644
--- a/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts
+++ b/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts
@@ -992,6 +992,67 @@ vdd-pins {
 			bias-pull-down;
 		};
 	};
+
+	hall {
+                hall_intr3: hall_intr3 {
+                    mux {
+                        pins = "gpio101";
+                        function = "gpio";
+                    };
+
+                    config {
+                        pins = "gpio101";
+                        drive-strength = <2>;
+                        bias-pull-up;
+                        input-enable;
+                    };
+                };
+
+                hall_intr4: hall_intr4 {
+                    mux {
+                        pins = "gpio134";
+                        function = "gpio";
+                    };
+
+                    config {
+                        pins = "gpio134";
+                        drive-strength = <2>;
+                        bias-pull-up;
+                        input-enable;
+                    };
+                };
+        };
+
+        idt {
+                idt_int: idt_int {
+                    mux {
+                        pins = "gpio95";
+                        function = "gpio";
+                    };
+
+                    config {
+                        pins = "gpio95";
+                        drive-strength = <2>;
+                        bias-pull-up;
+                        input-enable;
+                    };
+                };
+
+                txon_enable: txon_enable {
+                    mux {
+                        pins = "gpio3";
+                        function = "gpio";
+                    };
+
+                    config {
+                        pins = "gpio3";
+                        drive-strength = <16>;
+                        bias-disable;
+                        bias-pull-down;
+                        output-low;
+                    };
+                };
+        };
 };
 
 &uart2 {
@@ -1131,4 +1192,19 @@ &pm8150_sdam_2 {
        rtc_offset: rtc-offset@bc {
                reg = <0xbc 0x4>;
        };
-};
\ No newline at end of file
+};
+
+&i2c17 {
+	status = "okay";
+
+        idtp9418: idtp9418@3b {
+                compatible = "idt,p9418";
+                reg = <0x3b>;
+                status = "okay";
+                irq-gpios = <&tlmm 95 GPIO_ACTIVE_HIGH>;
+		pen-det-active-high-gpios = <&tlmm 101 GPIO_ACTIVE_HIGH>;
+		pen-det-active-low-gpios =<&tlmm 134 GPIO_ACTIVE_HIGH>;
+                enable-gpios = <&tlmm 3 GPIO_ACTIVE_HIGH>;
+                boost-enable-gpios = <&tlmm 47 GPIO_ACTIVE_HIGH>;
+        };
+};
diff --git a/arch/arm64/configs/sm8150.config b/arch/arm64/configs/sm8150.config
index db7298f8341a..957e71b043b6 100644
--- a/arch/arm64/configs/sm8150.config
+++ b/arch/arm64/configs/sm8150.config
@@ -11,6 +11,7 @@ CONFIG_DRM_PANEL_NOVATEK_NT36523=m
 CONFIG_BACKLIGHT_KTZ8866=m
 CONFIG_TOUCHSCREEN_NT36523_SPI=m
 CONFIG_SND_SOC_CS35L41_I2C=m
+CONFIG_CHARGER_IDTP9418=m
 
 # Xiaomi POCO X3 Pro
 CONFIG_SPI_XIAOMI_TP=y
diff --git a/drivers/power/supply/Kconfig b/drivers/power/supply/Kconfig
index 0ed12a5a30c8..d80432fcec20 100644
--- a/drivers/power/supply/Kconfig
+++ b/drivers/power/supply/Kconfig
@@ -1082,4 +1082,14 @@ config FUEL_GAUGE_MM8013
 	  the state of charge, temperature, cycle count, actual and design
 	  capacity, etc.
 
+config CHARGER_IDTP9418
+	tristate "idtp9418 Wireless Power Charger driver"
+	depends on I2C
+	select REGMAP_I2C
+	help
+	  Say Y here to enable the idtp9418 Wireless Power Charger driver.
+	  It enables the control and monitoring of the reverse wireless power
+	  charging in used in some Xiaomi tablet devices for recharging
+	  the stylus.
+
 endif # POWER_SUPPLY
diff --git a/drivers/power/supply/Makefile b/drivers/power/supply/Makefile
index 4f00f8e82c79..ca1411fd7cee 100644
--- a/drivers/power/supply/Makefile
+++ b/drivers/power/supply/Makefile
@@ -122,4 +122,5 @@ obj-$(CONFIG_CHARGER_SURFACE)	+= surface_charger.o
 obj-$(CONFIG_BATTERY_UG3105)	+= ug3105_battery.o
 obj-$(CONFIG_BATTERY_QCOM_FG)	+= qcom_fg.o
 obj-$(CONFIG_CHARGER_QCOM_SMB2)	+= qcom_smbx.o
-obj-$(CONFIG_FUEL_GAUGE_MM8013)	+= mm8013.o
\ No newline at end of file
+obj-$(CONFIG_FUEL_GAUGE_MM8013)	+= mm8013.o
+obj-$(CONFIG_CHARGER_IDTP9418)	+= idtp9418.o
\ No newline at end of file
diff --git a/drivers/power/supply/idtp9418.c b/drivers/power/supply/idtp9418.c
new file mode 100644
index 000000000000..48a62d594ef4
--- /dev/null
+++ b/drivers/power/supply/idtp9418.c
@@ -0,0 +1,794 @@
+// SPDX-License-Identifier: GPL-2.0
+/****************************************************************
+ *
+ *     idtp9418.c - idtp9418 Wireless Power Charger Monitor driver
+ *
+ *     (c) 2021 XiaoMi, Inc.
+ *     (c) 2025 Viola Guerrera
+ *
+ *     Original author: bsp-open <bsp-open@xiaomi.com>
+ *     Adapted and cleaned up by: Viola Guerrera <viooo@anche.no>
+ *     Tested on xiaomi-nabu
+ *
+ ****************************************************************/
+
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/sched.h>
+#include <linux/i2c.h>
+#include <linux/regmap.h>
+#include <linux/power_supply.h>
+#include "idtp9418.h"
+
+#define LIMIT_SOC 85
+#define CHARGE_MONITOR_INTERVAL 2 * HZ
+
+struct idtp9418_device_info;
+
+struct idtp9418_access_func
+{
+	int (*read)(struct idtp9418_device_info *di, uint16_t reg, uint8_t *val);
+	int (*write)(struct idtp9418_device_info *di, uint16_t reg, uint8_t val);
+	int (*read_buf)(struct idtp9418_device_info *di, uint16_t reg, uint8_t *buf,
+					uint32_t size);
+	int (*write_buf)(struct idtp9418_device_info *di, uint16_t reg, uint8_t *buf,
+					 uint32_t size);
+};
+
+struct idtp9418_device_info
+{
+	int chip_enable;
+	char *name;
+	struct power_supply *psy;
+	int charge_limit;
+	bool is_charging;
+	struct device *dev;
+	struct idtp9418_access_func bus;
+	struct regmap *regmap;
+	struct
+	{
+		struct gpio_desc *enable;
+		struct gpio_desc *boost_enable;
+		struct gpio_desc *idt_irq;
+		struct gpio_desc *pen_det_active_high;
+		struct gpio_desc *pen_det_active_low;
+	} gpios;
+	struct delayed_work irq_work;
+	struct delayed_work hall_irq_work;
+	struct delayed_work charge_monitor_work;
+	struct mutex i2c_lock;
+};
+
+static enum power_supply_property idtp9418_props[] = {
+	POWER_SUPPLY_PROP_CAPACITY,
+	POWER_SUPPLY_PROP_CHARGE_CONTROL_END_THRESHOLD,
+	POWER_SUPPLY_PROP_ONLINE,
+};
+
+void reverse_clrInt(struct idtp9418_device_info *di);
+
+static int idtp9418_read(struct idtp9418_device_info *di, uint16_t reg, uint8_t *val)
+{
+	unsigned int temp;
+	int rc;
+	msleep(100);
+
+	mutex_lock(&di->i2c_lock);
+	rc = regmap_read(di->regmap, reg, &temp);
+	if (rc >= 0)
+		*val = (u8)temp;
+
+	mutex_unlock(&di->i2c_lock);
+	if (rc < 0)
+	{
+		dev_err(di->dev, "read error: %d. while reading reg: 0x%04x val: 0x%02x\n",
+				rc, reg, *val);
+		return rc;
+	}
+	return rc;
+}
+
+static int idtp9418_write(struct idtp9418_device_info *di, uint16_t reg, uint8_t val)
+{
+	int rc = 0;
+
+	mutex_lock(&di->i2c_lock);
+	rc = regmap_write(di->regmap, reg, val);
+	if (rc < 0)
+	{
+		dev_err(di->dev, "write error: %d at reg: 0x%04x val: 0x%02x\n",
+				rc, reg, val);
+	}
+	mutex_unlock(&di->i2c_lock);
+	return rc;
+}
+
+static int idtp9418_read_buffer(struct idtp9418_device_info *di, uint16_t reg, uint8_t *buf,
+								uint32_t size)
+{
+	int rc = 0;
+
+	while (size--)
+	{
+		rc = di->bus.read(di, reg++, buf++);
+		if (rc < 0)
+		{
+			dev_err(di->dev, "[idt] read buf error: %d\n", rc);
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+static int idtp9418_write_buffer(struct idtp9418_device_info *di, uint16_t reg, uint8_t *buf,
+								 uint32_t size)
+{
+	int rc = 0;
+
+	while (size--)
+	{
+		rc = di->bus.write(di, reg++, *buf++);
+		if (rc < 0)
+		{
+			dev_err(di->dev, "[idt] write error: %d\n", rc);
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+/*
+	Turn on/off the power to the chip
+	This turn on "reverse-enable" gpio which contorls the
+	mosfet to power the TPS22916 boost converter
+	By default, vout of the TPS22916 is the same as the input voltage
+*/
+static void idtp9418_set_chip_power(struct idtp9418_device_info *di,
+									int enable)
+{
+	gpiod_set_value(di->gpios.enable, !!enable);
+}
+
+static void idt_set_reverse_fod(struct idtp9418_device_info *di, int mw)
+{
+	uint8_t mw_l, mw_h;
+	if (!di)
+		return;
+	mw_l = mw & 0xff;
+	mw_h = mw >> 8;
+	di->bus.write(di, REG_FOD_LOW, mw_l);
+	di->bus.write(di, REG_FOD_HIGH, mw_h);
+	dev_info(di->dev, "set reverse fod: %d\n", mw);
+}
+
+/*
+	Use the reverse-boost-enable-gpio gpio to turn on/off the voltage boosting.
+	This needs to be turned on after the chip is powered.
+*/
+static void idtp9418_enable_boost_converter(struct idtp9418_device_info *di, int enable)
+{
+	gpiod_set_value(di->gpios.boost_enable, !!enable);
+}
+
+static ssize_t chip_version_show(struct device *dev,
+								 struct device_attribute *attr, char *buf)
+{
+	uint8_t chip_id_l, chip_id_h, chip_rev, cust_id, status, vset;
+	uint8_t fw_otp_ver[4], fw_app_ver[4];
+	struct i2c_client *client = container_of(dev, struct i2c_client, dev);
+	struct idtp9418_device_info *di = i2c_get_clientdata(client);
+	msleep(100);
+	di->bus.read(di, REG_STATUS_L, &status);
+	di->bus.read(di, REG_VOUT_SET, &vset);
+
+	di->bus.read(di, REG_CHIP_ID_L, &chip_id_l);
+	di->bus.read(di, REG_CHIP_ID_H, &chip_id_h);
+	di->bus.read(di, REG_CHIP_REV, &chip_rev);
+	chip_rev = chip_rev >> 4;
+	di->bus.read(di, REG_CTM_ID, &cust_id);
+	di->bus.read_buf(di, REG_OTPFWVER_ADDR, fw_otp_ver, 4);
+	di->bus.read_buf(di, REG_EPRFWVER_ADDR, fw_app_ver, 4);
+
+	return snprintf(
+		buf, PAGE_SIZE,
+		"chip_id_l:%02x\nchip_id_h:%02x\nchip_rev:%02x\ncust_id:%02x status:%02x vset:%02x\n otp_ver:%x.%x.%x.%x\n app_ver:%x.%x.%x.%x\n",
+		chip_id_l, chip_id_h, chip_rev, cust_id, status, vset,
+		fw_otp_ver[0], fw_otp_ver[1], fw_otp_ver[2],
+		fw_otp_ver[3], fw_app_ver[0], fw_app_ver[1],
+		fw_app_ver[2], fw_app_ver[3]);
+}
+
+static int idt_get_reverse_vout(struct idtp9418_device_info *di)
+{
+	u8 vout_l, vout_h;
+	int vout = -1;
+
+	if (!di)
+		return vout;
+
+	di->bus.read(di, REG_REVERSE_VIN_LOW, &vout_l);
+	di->bus.read(di, REG_REVERSE_VIN_HIGH, &vout_h);
+	vout = vout_l | (vout_h << 8);
+	dev_info(di->dev, "[reverse] vout is %d\n", vout);
+	return vout;
+}
+
+static int idt_get_reverse_soc(struct idtp9418_device_info *di)
+{
+	u8 soc = -1;
+
+	if (!di)
+		return soc;
+
+	di->bus.read(di, REG_CHG_STATUS, &soc);
+	if ((soc < 0) || (soc > 0x64))
+	{
+		if (soc == 0xFF)
+		{
+			dev_info(di->dev, "[reverse] soc is default 0xFF\n");
+			return -1;
+		}
+		else
+		{
+			dev_info(di->dev, "[reverse] soc illegal: %d\n", soc);
+			return -1;
+		}
+	}
+	return soc;
+}
+
+static int idt_get_reverse_iin(struct idtp9418_device_info *di)
+{
+	u8 iin_l, iin_h;
+	int iin = -1;
+
+	if (!di)
+		return iin;
+
+	di->bus.read(di, REG_REVERSE_IIN_LOW, &iin_l);
+	di->bus.read(di, REG_REVERSE_IIN_HIGH, &iin_h);
+	iin = iin_l | (iin_h << 8);
+	dev_info(di->dev, "[reverse] iin is %d\n", iin);
+	return iin;
+}
+
+static int idt_get_reverse_temp(struct idtp9418_device_info *di)
+{
+	u8 temp = -1;
+
+	if (!di)
+		return temp;
+
+	di->bus.read(di, REG_REVERSE_TEMP, &temp);
+	dev_info(di->dev, "[reverse] temp is %d\n", temp);
+	return temp;
+}
+
+static ssize_t chip_status_show(struct device *dev, struct device_attribute *attr,
+								char *buf)
+{
+	struct i2c_client *client = container_of(dev, struct i2c_client, dev);
+	struct idtp9418_device_info *di = i2c_get_clientdata(client);
+	int vout, iin, temp, soc;
+
+	vout = idt_get_reverse_vout(di);
+	iin = idt_get_reverse_iin(di);
+	temp = idt_get_reverse_temp(di);
+	soc = idt_get_reverse_soc(di);
+
+	return snprintf(buf, PAGE_SIZE, "vout:%d\niin:%d\ntemp:%d\nsoc:%d\n",
+					vout, iin, temp, soc);
+}
+
+static ssize_t chip_powered_show(struct device *dev,
+								 struct device_attribute *attr, char *buf)
+{
+	struct i2c_client *client = container_of(dev, struct i2c_client, dev);
+	struct idtp9418_device_info *di = i2c_get_clientdata(client);
+
+	return snprintf(buf, PAGE_SIZE, "%d\n", di->chip_enable);
+}
+
+static ssize_t chip_power_store(struct device *dev,
+								struct device_attribute *attr, const char *buf, size_t count)
+{
+	bool enable;
+	int ret;
+	struct i2c_client *client = container_of(dev, struct i2c_client, dev);
+	struct idtp9418_device_info *di = i2c_get_clientdata(client);
+
+	ret = kstrtobool(buf, &enable);
+	if (ret < 0)
+		return ret;
+
+	idtp9418_set_chip_power(di, enable);
+
+	return count;
+}
+
+static DEVICE_ATTR(chip_version, S_IRUGO, chip_version_show, NULL);
+static DEVICE_ATTR(chip_powered, S_IWUSR | S_IRUGO, chip_powered_show, chip_power_store);
+static DEVICE_ATTR(chip_status, S_IRUGO, chip_status_show, NULL);
+
+static struct attribute *sysfs_attrs[] = {
+	&dev_attr_chip_version.attr,
+	&dev_attr_chip_powered.attr,
+	&dev_attr_chip_status.attr,
+	NULL,
+};
+
+static const struct attribute_group sysfs_group_attrs = {
+	.attrs = sysfs_attrs,
+};
+
+static irqreturn_t idtp9418_irq_handler(int irq, void *dev_id)
+{
+	struct idtp9418_device_info *di = dev_id;
+	pm_stay_awake(di->dev);
+	schedule_delayed_work(&di->irq_work, msecs_to_jiffies(10));
+	return IRQ_HANDLED;
+}
+
+static irqreturn_t idtp9418_hall_irq_handler(int irq, void *dev_id)
+{
+	struct idtp9418_device_info *di = dev_id;
+	pm_stay_awake(di->dev);
+	schedule_delayed_work(&di->hall_irq_work, msecs_to_jiffies(100));
+
+	return IRQ_HANDLED;
+}
+
+static int idtp9418_gpio_init(struct idtp9418_device_info *di)
+{
+	struct
+	{
+		const char *name;
+		struct gpio_desc **gpio;
+		enum gpiod_flags flags;
+	} gpio_map[] = {
+		{"irq", &di->gpios.idt_irq, GPIOD_IN},
+		{"enable", &di->gpios.enable, GPIOD_OUT_HIGH},
+		{"boost-enable", &di->gpios.boost_enable, GPIOD_OUT_HIGH},
+		{"pen-det-active-high", &di->gpios.pen_det_active_high, GPIOD_IN},
+		{"pen-det-active-low", &di->gpios.pen_det_active_low, GPIOD_IN}};
+
+	for (int i = 0; i < ARRAY_SIZE(gpio_map); i++)
+	{
+		*gpio_map[i].gpio = devm_gpiod_get(di->dev, gpio_map[i].name, gpio_map[i].flags);
+		if (IS_ERR(*gpio_map[i].gpio))
+		{
+			dev_err(di->dev, "Unable to get GPIO: %s\n", gpio_map[i].name);
+			return PTR_ERR(*gpio_map[i].gpio);
+		}
+	}
+
+	struct
+	{
+		const char *name;
+		struct gpio_desc *gpio;
+		irq_handler_t handler;
+		unsigned long irqflags;
+	} irq_map[] = {
+		{
+			"idt-irq",
+			di->gpios.idt_irq,
+			idtp9418_irq_handler,
+			IRQF_TRIGGER_FALLING,
+		},
+		{
+			"hall-pen-l-irq",
+			di->gpios.pen_det_active_low,
+			idtp9418_hall_irq_handler,
+			IRQF_TRIGGER_FALLING | IRQF_TRIGGER_RISING,
+		},
+		{
+			"hall-pen-h-irq",
+			di->gpios.pen_det_active_high,
+			idtp9418_hall_irq_handler,
+			IRQF_TRIGGER_FALLING | IRQF_TRIGGER_RISING,
+		}};
+	for (int i = 0; i < ARRAY_SIZE(irq_map); i++)
+	{
+		int ret = devm_request_irq(di->dev, gpiod_to_irq(irq_map[i].gpio),
+								   irq_map[i].handler,
+								   irq_map[i].irqflags,
+								   irq_map[i].name, di);
+		if (ret)
+		{
+			dev_err(di->dev, "failed to request gpio \"%s\" IRQ\n", irq_map[i].name);
+		}
+	}
+
+	return 0;
+}
+
+static bool reverse_need_irq_cleared(struct idtp9418_device_info *di, uint32_t val)
+{
+	uint8_t int_buf[4];
+	uint32_t int_val;
+	int rc = -1;
+
+	rc = di->bus.read_buf(di, REG_SYS_INT, int_buf, 4);
+	if (rc < 0)
+	{
+		dev_err(di->dev, "%s: read int state error\n", __func__);
+		return false;
+	}
+	int_val = int_buf[0] | (int_buf[1] << 8) | (int_buf[2] << 16) |
+			  (int_buf[3] << 24);
+	if (int_val && (int_val == val))
+	{
+		dev_info(di->dev, "irq clear wrong, retry: 0x%08x\n", int_val);
+		return true;
+	}
+
+	return false;
+}
+
+static void idtp9418_reverse_charge_enable(struct idtp9418_device_info *di)
+{
+	u8 mode = 0;
+	int i = 0;
+	int fod_set = REVERSE_FOD;
+
+	/* set reverse fod to REVERSE_FOD */
+	idt_set_reverse_fod(di, fod_set);
+	for (i = 0; i < 3; i++)
+	{
+		di->bus.write(di, REG_TX_CMD, TX_EN | TX_FOD_EN);
+		msleep(50);
+		di->bus.read(di, REG_TX_DATA, &mode);
+		dev_info(di->dev, "tx data(0078): 0x%x\n", mode);
+		if (mode & BIT(0))
+		{
+			dev_info(di->dev, "start reverse charging\n");
+			break;
+		}
+		else
+		{
+			dev_err(di->dev, "idt set reverse charge failed, retry: %d\n", i);
+		}
+	}
+
+	if (i < 3)
+	{
+		dev_info(di->dev, "reverse charging start success\n");
+		di->is_charging = true;
+		// Start timer
+		schedule_delayed_work(&di->charge_monitor_work,
+							  msecs_to_jiffies(CHARGE_MONITOR_INTERVAL));
+	}
+	else
+	{
+		// Clean up
+		dev_info(di->dev, "reverse charging failed start\n");
+		idtp9418_enable_boost_converter(di, false);
+		idtp9418_set_chip_power(di, false);
+		di->is_charging = false;
+		cancel_delayed_work_sync(&di->charge_monitor_work);
+	}
+}
+
+static void idtp9418_charge_monitor_work(struct work_struct *work)
+{
+	struct idtp9418_device_info *di =
+		container_of(work, struct idtp9418_device_info, charge_monitor_work.work);
+
+	int soc;
+	bool attached;
+
+	soc = idt_get_reverse_soc(di);
+	attached = (!gpiod_get_value(di->gpios.pen_det_active_low) &&
+				gpiod_get_value(di->gpios.pen_det_active_high));
+
+	dev_info(di->dev, "[charge-monitor] pen=%d soc=%d\n", attached, soc);
+	power_supply_changed(di->psy);
+
+	if (!attached || soc < 0 || soc >= di->charge_limit)
+	{
+		dev_info(di->dev, "[charge-monitor] stopping reverse charging. attached=%d soc=%d\n",
+				 attached, soc);
+
+		// Turn off charging and power
+		idtp9418_enable_boost_converter(di, false);
+		idtp9418_set_chip_power(di, false);
+		di->is_charging = false;
+
+		return;
+	}
+
+	schedule_delayed_work(&di->charge_monitor_work,
+						  msecs_to_jiffies(CHARGE_MONITOR_INTERVAL));
+}
+
+static void idtp9418_irq_hall_work(struct work_struct *work)
+{
+	struct idtp9418_device_info *di = container_of(work, struct idtp9418_device_info, hall_irq_work.work);
+	bool attached = false;
+
+	if (!gpiod_get_value(di->gpios.pen_det_active_low) && gpiod_get_value(di->gpios.pen_det_active_high))
+	{
+		dev_info(di->dev, "[idtp] Pen attached! Turning on reverse charging... \n");
+		attached = true;
+	}
+	else
+	{
+		dev_info(di->dev, "[idtp] Pen detached! Turning off reverse charging... \n");
+		di->is_charging = false;
+		cancel_delayed_work_sync(&di->charge_monitor_work);
+	}
+
+	idtp9418_set_chip_power(di, false);
+	idtp9418_enable_boost_converter(di, false);
+	msleep(100);
+	if (attached)
+	{
+		idtp9418_enable_boost_converter(di, true);
+		msleep(600);
+		idtp9418_set_chip_power(di, true);
+		msleep(3000);
+		idtp9418_reverse_charge_enable(di);
+	}
+	pm_relax(di->dev);
+	power_supply_changed(di->psy);
+}
+
+static void idtp9418_irq_work(struct work_struct *work)
+{
+	struct idtp9418_device_info *di = container_of(work, struct idtp9418_device_info, irq_work.work);
+
+	dev_err(di->dev, "%s: Entered irq work\n", __func__);
+
+	u8 int_buf[4] = {0};
+	u32 int_val = 0;
+	int rc = 0;
+	int irq_level;
+
+	if (gpiod_get_value(di->gpios.idt_irq))
+	{
+		dev_info(di->dev, "irq is high level, ignore%d\n", irq_level);
+		pm_relax(di->dev);
+		return;
+	}
+
+	rc = di->bus.read_buf(di, REG_SYS_INT, int_buf, 4);
+	if (rc < 0)
+	{
+		dev_err(di->dev, "[idt]read int state error: %d\n", rc);
+	}
+	int_val =
+		int_buf[0] | (int_buf[1] << 8) | (int_buf[2] << 16) | (int_buf[3] << 24);
+
+	dev_info(di->dev, "Irq got: %d\n", int_val);
+
+	msleep(5);
+	if (reverse_need_irq_cleared(di, int_val))
+	{
+		reverse_clrInt(di);
+		msleep(5);
+	}
+	pm_relax(di->dev);
+}
+
+static struct regmap_config i2c_idtp9418_regmap_config = {
+	.reg_bits = 16,
+	.val_bits = 8,
+	.max_register = 0xFFFF,
+};
+
+static const struct i2c_device_id idtp9418_id[] = {
+	{IDT_DRIVER_NAME, 0},
+	{},
+};
+
+static const struct of_device_id idt_match_table[] = {{.compatible =
+														   "idt,p9418"},
+													  {}};
+
+MODULE_DEVICE_TABLE(i2c, idtp9418_id);
+
+void reverse_clrInt(struct idtp9418_device_info *di)
+{
+	uint8_t clr_buf[4] = {0xFF, 0xFF, 0xFF, 0xFF};
+	di->bus.write_buf(di, REG_SYS_INT_CLR, clr_buf, sizeof(clr_buf));
+	di->bus.write(di, REG_TX_CMD, TX_FOD_EN | TX_CLRINT);
+
+	dev_err(di->dev, "Interrupt cleared!\n");
+}
+
+static int idtp9418_get_property(struct power_supply *psy,
+								 enum power_supply_property psp,
+								 union power_supply_propval *val)
+{
+	struct idtp9418_device_info *di = power_supply_get_drvdata(psy);
+
+	switch (psp)
+	{
+	case POWER_SUPPLY_PROP_CAPACITY:
+		if (!di->is_charging)
+			return 0;
+		val->intval = idt_get_reverse_soc(di);
+		if (val->intval < 0)
+			val->intval = 0;
+		return 0;
+
+	case POWER_SUPPLY_PROP_CHARGE_CONTROL_END_THRESHOLD:
+		val->intval = di->charge_limit;
+		return 0;
+
+	case POWER_SUPPLY_PROP_ONLINE:
+		val->intval = di->is_charging ? 1 : 0;
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+static int idtp9418_property_is_writeable(struct power_supply *psy,
+										  enum power_supply_property psp)
+{
+	switch (psp)
+	{
+	case POWER_SUPPLY_PROP_CHARGE_CONTROL_END_THRESHOLD:
+		return 1;
+	default:
+		return 0;
+	}
+}
+
+static int idtp9418_set_property(struct power_supply *psy,
+								 enum power_supply_property psp,
+								 const union power_supply_propval *val)
+{
+	struct idtp9418_device_info *di = power_supply_get_drvdata(psy);
+
+	switch (psp)
+	{
+	case POWER_SUPPLY_PROP_CHARGE_CONTROL_END_THRESHOLD:
+		if (val->intval < 0 || val->intval > 95)
+			return -EINVAL;
+
+		di->charge_limit = val->intval;
+		dev_info(di->dev, "Charge limit set to %d%%\n", di->charge_limit);
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+static int idtp9418_probe(struct i2c_client *client)
+{
+	int ret;
+	struct idtp9418_device_info *di;
+	struct i2c_adapter *adapter = to_i2c_adapter(client->dev.parent);
+	struct power_supply_config idtp_cfg;
+	struct power_supply_desc *psy_desc;
+
+	if (!i2c_check_functionality(adapter, I2C_FUNC_SMBUS_BYTE))
+	{
+		dev_err(&client->dev, "i2c check functionality failed!\n");
+		return -EIO;
+	}
+
+	di = devm_kzalloc(&client->dev, sizeof(*di), GFP_KERNEL);
+	if (!di)
+	{
+		dev_err(&client->dev,
+				"i2c allocated device info data failed!\n");
+		return -ENOMEM;
+	}
+
+	di->name = IDT_DRIVER_NAME;
+	di->dev = &client->dev;
+	di->chip_enable = 0;
+
+	di->is_charging = false;
+	di->charge_limit = LIMIT_SOC;
+
+	di->regmap = devm_regmap_init_i2c(client, &i2c_idtp9418_regmap_config);
+	if (!di->regmap)
+		return -ENODEV;
+	di->bus.read = idtp9418_read;
+	di->bus.write = idtp9418_write;
+	di->bus.read_buf = idtp9418_read_buffer;
+	di->bus.write_buf = idtp9418_write_buffer;
+
+	mutex_init(&di->i2c_lock);
+	device_init_wakeup(&client->dev, true);
+	i2c_set_clientdata(client, di);
+
+	ret = idtp9418_gpio_init(di);
+	if (ret < 0)
+	{
+		dev_err(di->dev, "%s: gpio init error [%d]\n", __func__, ret);
+		return ret;
+	}
+
+	if (sysfs_create_group(&client->dev.kobj, &sysfs_group_attrs))
+	{
+		dev_err(&client->dev, "create sysfs attrs failed!\n");
+		return -EIO;
+	}
+
+	idtp_cfg.drv_data = di;
+
+	INIT_DELAYED_WORK(&di->irq_work, idtp9418_irq_work);
+	INIT_DELAYED_WORK(&di->hall_irq_work, idtp9418_irq_hall_work);
+	INIT_DELAYED_WORK(&di->charge_monitor_work, idtp9418_charge_monitor_work);
+
+	idtp9418_set_chip_power(di, false);
+
+	// Initialize psy
+
+	psy_desc = devm_kzalloc(&client->dev, sizeof(*psy_desc), GFP_KERNEL);
+	if (!psy_desc)
+		return -ENOMEM;
+
+	psy_desc->name = "idtp9418";
+	psy_desc->type = POWER_SUPPLY_TYPE_WIRELESS;
+	psy_desc->properties = idtp9418_props;
+	psy_desc->num_properties = ARRAY_SIZE(idtp9418_props);
+	psy_desc->get_property = idtp9418_get_property;
+	psy_desc->set_property = idtp9418_set_property;
+	psy_desc->property_is_writeable = idtp9418_property_is_writeable;
+
+	idtp_cfg.drv_data = di;
+
+	di->psy = devm_power_supply_register(&client->dev, psy_desc, &idtp_cfg);
+	if (IS_ERR(di->psy))
+	{
+		dev_err(&client->dev, "Failed to register power supply\n");
+		return PTR_ERR(di->psy);
+	}
+
+	return 0;
+}
+
+static void idtp9418_remove(struct i2c_client *client)
+{
+	sysfs_remove_group(&client->dev.kobj, &sysfs_group_attrs);
+
+	struct idtp9418_device_info *di = i2c_get_clientdata(client);
+	i2c_set_clientdata(client, NULL);
+
+	cancel_delayed_work_sync(&di->irq_work);
+	cancel_delayed_work_sync(&di->hall_irq_work);
+	cancel_delayed_work_sync(&di->charge_monitor_work);
+}
+
+static void idtp9418_shutdown(struct i2c_client *client)
+{
+}
+
+static struct i2c_driver idtp9418_driver = {
+	.driver = {
+		.name = IDT_DRIVER_NAME,
+		.owner = THIS_MODULE,
+		.of_match_table = idt_match_table,
+	},
+	.probe = idtp9418_probe,
+	.remove = idtp9418_remove,
+	.shutdown = idtp9418_shutdown,
+	.id_table = idtp9418_id,
+};
+
+static int __init idt_init(void)
+{
+	return i2c_add_driver(&idtp9418_driver);
+}
+
+static void __exit idt_exit(void)
+{
+	i2c_del_driver(&idtp9418_driver);
+}
+
+module_init(idt_init);
+module_exit(idt_exit);
+
+MODULE_AUTHOR("Viola Guerrera <viooo@anche.no>");
+MODULE_DESCRIPTION("idtp9418 Wireless Power Charger Monitor driver");
+MODULE_LICENSE("GPL v2");
diff --git a/drivers/power/supply/idtp9418.h b/drivers/power/supply/idtp9418.h
new file mode 100644
index 000000000000..9829e07f041f
--- /dev/null
+++ b/drivers/power/supply/idtp9418.h
@@ -0,0 +1,208 @@
+#ifndef __IDTP9148_H__
+#define __IDTP9148_H__
+
+#define IDT_DRIVER_NAME "idtp9418"
+#define IDT_I2C_ADDR 0x3b
+
+#define HSCLK 60000
+
+#define ADJUST_METE_MV 35
+#define IDTP9220_DELAY 2000
+#define CHARGING_FULL 100
+#define CHARGING_NEED 95
+
+/* status low regiter bits define */
+#define STATUS_VOUT_ON (1 << 7)
+#define STATUS_VOUT_OFF (1 << 6)
+#define STATUS_TX_DATA_RECV (1 << 4)
+#define STATUS_OV_TEMP (1 << 2)
+#define STATUS_OV_VOL (1 << 1)
+#define STATUS_OV_CURR (1 << 0)
+
+#define OVER_EVENT_OCCUR (STATUS_OV_TEMP | STATUS_OV_VOL | STATUS_OV_CURR)
+#define TOGGLE_LDO_ON_OFF (1 << 1)
+
+/* interrupt register bits define */
+#define INT_RXREADY (1 << 22)
+#define INT_POWER_LOSS (1 << 21)
+#define INT_OVERCURRENT_HIGH (1 << 20)
+#define RENEG_FAIL (1 << 19)
+#define RENEG_SUCCESS (1 << 18)
+#define INT_RPP_READ (1 << 17)
+#define INT_EXT5V_FAIL (1 << 16)
+#define INT_VSWITCH_FAIL (1 << 15)
+#define INT_SLEEPMODE (1 << 14)
+#define INT_IDAUTH_SUCESS (1 << 13)
+#define INT_IDAUTH_FAIL (1 << 12)
+#define INT_SEND_SUCESS (1 << 11)
+#define INT_SEND_TIMEOUT (1 << 10)
+#define INT_AUTH_SUCESS (1 << 9)
+#define INT_AUTH_FAIL (1 << 8)
+#define INT_VOUT_OFF (1 << 7)
+#define INT_VOUT_ON (1 << 6)
+#define INT_MODE_CHANGE (1 << 5)
+#define INT_TX_DATA_RECV (1 << 4)
+#define INT_VSWITCH_SUCESS (1 << 3)
+#define INT_OV_TEMP (1 << 2)
+#define INT_OV_VOL (1 << 1)
+#define INT_OV_CURR (1 << 0)
+
+/*TRX int bits define*/
+#define INT_GET_BLE_ADDR (1 << 8)
+#define INT_INIT_TX (1 << 7)
+#define INT_GET_DPING (1 << 6)
+#define INT_GET_PPP (1 << 5)
+#define INT_GET_CFG (1 << 4)
+#define INT_GET_ID (1 << 3)
+#define INT_GET_SS (1 << 2)
+#define INT_START_DPING (1 << 1)
+#define INT_EPT_TYPE (1 << 0)
+
+#define TX_TOGGLE BIT(6)   // toggle work mode(148K or 190K)
+#define TX_FOD_EN BIT(5)   // enable FOD
+#define TX_WD BIT(4)       // enable WD
+#define TX_SEND_FSK BIT(3) // SEND_FSK
+#define TX_DIS BIT(2)      // disable tx
+#define TX_CLRINT BIT(1)   // clr int
+#define TX_EN BIT(0)       // enable tx mode
+
+/* used registers define */
+#define REG_CHIP_ID_L 0x0000
+#define REG_CHIP_ID_H 0x0001
+#define REG_CHIP_REV 0x0002
+#define REG_CTM_ID 0x0003
+#define REG_OTPFWVER_ADDR 0x0004 // OTP firmware version
+#define REG_EPRFWVER_ADDR 0x001c // EEPROM firmware version
+#define REG_SYS_INT_CLR 0x0028   // int clear
+#define REG_SYS_INT 0x0030       // interrupt
+#define REG_STATUS_L 0x0034
+#define REG_STATUS_H 0x0035
+#define REG_INTR_L 0x0036
+#define REG_INTR_H 0x0037
+#define REG_INTR_EN_L 0x0038
+#define REG_INTR_EN_H 0x0039
+#define REG_CHG_STATUS 0x003A
+#define REG_ADC_VOUT_L 0x003C
+#define REG_ADC_VOUT_H 0x003D
+#define REG_VOUT_SET 0x003E
+#define REG_VRECT_ADJ 0x003F
+#define REG_ADC_VRECT 0x0040
+#define REG_RX_LOUT_L 0x0044
+#define REG_RX_LOUT_H 0x0045
+#define REG_RX_DIE_TEMP 0x0046
+#define REG_FREQ_ADDR 0x0048 // Operating Frequency, Fop(KHz) = 64 * 6000 /value * 256)
+#define REG_ILIM_SET 0x004A
+#define REG_SIGNAL_STRENGTH 0x004B
+#define REG_WPC_MODE 0x004D
+#define REG_SSCMND 0x004e // Command Register, COM (0x4E)
+#define REG_RX_RESET 0x004F
+#define REG_PROPPKT 0x0050      // Proprietary Packet Header Register, PPP_Header (0x50)
+#define REG_PPPDATA 0x0051      // PPP Data Value Register(0X51, 0x52, 0x53, 0x54, 0x55)
+#define REG_SSINTCLR 0x0056     // Interrupt Clear Registers, INT_Clear_L (0x56)
+#define REG_BCHEADER 0x0058     // Back Channel Packet Register (0x58)
+#define REG_BCDATA 0x0059       // Back Channel Packet Register (0x59, 0x5A, 0x5B, 0x5C)
+#define REG_FC_VOLTAGE_L 0x0078 // Fast Charging Voltage Register
+#define REG_FC_VOLTAGE_H 0x0079
+#define REG_REGULATOR_L 0x000C
+#define REG_REGULATOR_H 0x000d
+#define REG_POWER_MAX 0x0084 // Get the TX power on EPP mode.
+#define REG_TX_TYPE 0x00A2   // Get the TX type.
+#define REG_BLE_FLAG 0x00A4  // Get the rx ble flag.
+#define REG_CEP 0x00A5       // Get the CEP.
+#define REG_RPP 0x00A6       // Get the RPP.
+#define REG_MAX_POWER 0x008E
+#define REG_OCP_CONFIG 0x00F1
+#define REG_TEMPTER 0x00F2
+
+/*add for reverse charge*/
+#define REG_EPT_TYPE 0x0074
+#define REG_TX_CMD 0x0076
+#define REG_TX_DATA 0x0078
+#define REG_WROK_MODE 0x007B
+#define REG_VRECT_TAEGET 0x0090
+#define REG_FOD_LOW 0x0092
+#define REG_FOD_HIGH 0x0093
+#define REG_MAC_ADDR 0x00be
+
+#define REG_REVERSE_IIN_LOW 0x006e
+#define REG_REVERSE_IIN_HIGH 0x006f
+#define REG_REVERSE_VIN_LOW 0x0070
+#define REG_REVERSE_VIN_HIGH 0x0071
+#define REG_REVERSE_TEMP 0x007a
+
+#define EPT_POCP BIT(15)
+#define EPT_OTP BIT(14)
+#define EPT_FOD BIT(13)
+#define EPT_LVP BIT(12)
+#define EPT_OVP BIT(11)
+#define EPT_OCP BIT(10)
+#define EPT_RPP_TIMEOUT BIT(9)
+#define EPT_CEP_TIMEOUT BIT(8)
+#define EPT_CMD BIT(0)
+
+#define REG_EXTERNAL_5V 0x00dc // external 5v enable
+
+// RX -> TX
+#define PROPRIETARY18 0x18
+#define PROPRIETARY28 0x28
+#define PROPRIETARY38 0x38
+#define PROPRIETARY48 0x48
+#define PROPRIETARY58 0x58
+
+// bitmap for customer command
+#define BC_NONE 0x00
+#define BC_SET_FREQ 0x03
+#define BC_GET_FREQ 0x04
+#define BC_READ_FW_VER 0x05
+#define BC_READ_Iin 0x06
+#define BC_READ_Vin 0x07
+#define BC_SET_Vin 0x0a
+
+#define BC_ADAPTER_TYPE 0x0b
+#define BC_RESET 0x0c
+#define BC_READ_I2C 0x0d
+#define BC_WRITE_I2C 0x0e
+#define BC_VI2C_INIT 0x10
+#define BC_RX_ID_AUTH 0x3b
+#define BC_TX_COMPATIBLE_HWID 0x3f
+#define BC_TX_HWID 0x4c
+#define BC_TX_TOGGLE 0xc4
+#define CMD_GET_BLEMAC_2_0 0xb6
+#define CMD_GET_BLEMAC_5_3 0xb7
+#define CMD_FACOTRY_REVERSE 0x30
+
+#define NORMAL_MODE 0x1
+#define TAPER_MODE 0x2
+#define FULL_MODE 0x3
+#define RECHG_MODE 0x4
+
+// fw version to update
+#define FW_VERSION 0x1
+
+// add for crc verify
+#define CRC_VERIFY_LOW 0xB5
+#define CRC_VERIFY_HIGH 0xA9
+
+// add for reverse fod
+#define REVERSE_FOD 500
+
+/*CLRPOWERGOOD BIT(10), upper byte is 0x8*/
+#define CLRPOWERGOOD 0x04
+// bitmap for status flags
+// 1: indicates a pending interrupt for LDO Vout state change – from OFF to ON
+#define VOUTCHANGED BIT(7) // Stat_Vout_ON
+// 1: indicates a pending interrupt for TX Data Received. (Change from “No Received Data” state to “Data Received” state)
+#define TXDATARCVD BIT(4) // TX Data Received
+
+// bitmap for SSCmnd register 0x4e
+#define VSWITCH BIT(7)
+// If AP sets this bit to "1" then IDTP9220 M0 clears the interrupt corresponding to the bit(s) which has a value of “1”
+#define CLRINT BIT(5) // Clear Interrupt
+// If AP sets this bit to "1" then IDTP9220 M0 toggles LDO output once (from on to off, or from off to on), and then sets this bit to “0”
+#define LDOTGL BIT(1) // Toggle LDO On/OFF
+// If AP sets this bit to “1” then IDTP9220 M0 sends the Proprietary Packet
+#define SENDPROPP BIT(0) //  SEND RX Data
+
+#define SEND_DEVICE_AUTH BIT(2)
+
+#endif
-- 
2.51.1

